### PR TITLE
Add support for Stripe Connect

### PR DESCRIPTION
**Pull request by lukeburden**
I've implemented managed accounts. This code is stable and in-use, now I'd like to get it back into this project.

It adds the basic ability to create managed accounts, transact on their behalf and transfer funds to external bank accounts.

@paltman I've obviously got some work to do on tests, but before finishing those up I'd like the greenlight from you on:
- general approach
- addition of stripe_account to various models
- improvements to the robustness of the webhook handling code
- the initial and dynamic forms for collecting Bank Account data
- changes to sync_customers (I think this really needs to be built upon as it's still only customers, sources, charges and invoices)

Let me know what you think!

Thanks,

Luke

### CODE REVIEW

File: pinax/stripe/models.py [hunk 1]
```diff
@@ -132,10 +168,13 @@ class Customer(StripeObject):
...
     def stripe_customer(self):
-        return stripe.Customer.retrieve(self.stripe_id)
+        return stripe.Customer.retrieve(
+            self.stripe_id,
+            stripe_account=self.stripe_account
+        )
 
     def __str__(self):
-        return str(self.user)
+        return unicode(self.user)
```
Comment (paltman) [inline pinax/stripe/models.py]: shouldn't need `unicode` here because of the `@python_2_unicode_compatible` class decorator.

File: .pep8 [hunk 2]
```diff
@@ -0,0 +1,5 @@
+[flake8]
```
Comment (paltman) [inline .pep8]: why is this file added? there is already a config for `flake8` in `tox.ini`.  I'm also not so quick to ignore all those other things.

File: pinax/stripe/actions/accounts.py [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+    """
+    Create an Account.
+
+    Args:
+        country: two letter country code for where the individual lives
+
+    Returns:
+        a pinax.stripe.models.Account object
+    """
+    kwargs['country'] = country
```
Comment (paltman) [inline pinax/stripe/actions/accounts.py]: I'd like to stay consistent with rest of code base and not use single-quotes, but instead use double-quotes.

File: .pep8 (hunk 2 above)
Comment (lukeburden) [inline .pep8]: Just a config file for pep8/flake8. I'll remove it.

**Comment by lukeburden** [comment]
Hola @paltman - I've updated as requested. Glad it's mostly good! Let me know if there are any other adjustments you'd like me to make.

Is your preference for everything to be rebased into a single commit for merge?

**Comment by iMerica** [comment]
I'm glad I found this. I was about to embark on the same work. A few questions:

> I've implemented managed accounts 

Per Stripe's [documentation](https://stripe.com/docs/connect/accounts), there are three Connect Account Types:
- Standard
- Express
- Custom.

Which of these types are supported by this PR?

File: pinax/stripe/models.py [hunk 4]
```diff
@@ -24,8 +24,16 @@ class Meta:
         abstract = True
 
 
+class AccountRelatedStripeObject(StripeObject):
```
Comment (iMerica) [inline pinax/stripe/models.py]: Would it be better to have two abstract classes instead of the hierarchy (mixin pattern) ? This way only specific models that need the field can use it. Example:

```
class Customer(StripeObject, AccountRelatedStripeObject):
    ...
```
If you think Stripe Connect's business logic touches every model, then it makes sense, but I'm not sure it does.

Comment (lukeburden) [inline pinax/stripe/models.py]: @iMerica - "specific models that need the field can use it" - this is the intent. Models that can be account specific (which isn't all of them - it can be redundant to make them all partitioned by account) inherit from the AccountRelatedStripeObject abstract model.

Let me know if I've missed your point.

**Comment by lukeburden** [comment]
@iMerica My use case required full customisation of the Connect experience - so the pieces are added with the ability to add, update, transfer to and payout to connected accounts via the API.

The other account types won't allow you to manipulate connected account data via the API. But I suppose some subset of the Account info is available and sync'd via webhooks?

You'll need to experiment!

**Comment by iMerica** [comment]
Okay cool thanks for the update. I'm not an expert or maintainer in this codebase, but your changes LGTM. Maybe just add a new "Connect" section to the docs describing a hypothetical use case that can be supported by these changes?

**Comment by blueyed** [comment]
This needs to be rebased / updated.
What is the status of this?
I'd be glad to help out getting this finished.

**Comment by lukeburden** [comment]
Hello - I'll rebase my changes in the next couple of days. Keen to get this pushed through, too.

@blueyed - one way you could help would be playing around a bit with the DynamicManagedAccountForm and InitialManagedAccountForm classes. I wrote them and have not had a chance to give them much of a try, as the project I'm working on is lower-volume for account owners and we won't be hitting the later stages of account verification in the near-term.

**Comment by blueyed** [comment]
@lukeburden 
> Keen to get this pushed through, too.

Awesome.

1. I think there are things that could be moved out of this PR into separate ones, starting with "Improving speed of syncing large data sets from stripe" already (and the followups to it).
I could move them out into separate PRs, if you don't have the time yourself at the moment.

2. We're not making use of the forms ourselves, so they are likely the last for me to look at anyway.

3. Just for reference: some discussion on Stripe Connect integration for dj-stripe: https://github.com/dj-stripe/dj-stripe/issues/97#issuecomment-125778458

**Comment by paltman** [comment]
I'm very excited for this feature. Thanks for hammering on it.

File: pinax/stripe/models.py [hunk 5]
```diff
@@ -253,7 +292,12 @@ def status(self):
 
     @property
     def stripe_invoice(self):
-        return stripe.Invoice.retrieve(self.stripe_id)
+        return stripe.Invoice.retrieve(
+            self.stripe_id,
+            stripe_account=(
+                self.customer.stripe_account if self.customer_id else None
```
Comment (blueyed) [inline pinax/stripe/models.py]: Better to use `hasattr(self, 'customer')` here / or a `try/except self._meta.model.customer.RelatedObjectDoesNotExist`?!

File: pinax/stripe/models.py [hunk 6]
```diff
@@ -298,8 +342,118 @@ class Charge(StripeObject):
...
+    transfers_enabled = models.BooleanField(default=False)
+
+    verification_disabled_reason = models.TextField(null=True, blank=True)
+    verification_due_by = models.DateTimeField(null=True, blank=True)
+    verification_timestamp = models.DateTimeField(null=True, blank=True)
+    verification_fields_needed = JSONField(null=True)
+
+    @property
+    def stripe_account(self):
+        return stripe.Account.retrieve(self.stripe_id)
```
Comment (blueyed) [inline pinax/stripe/models.py]: This should be a cached property, similar to 79424ed probably?!

File: pinax/stripe/models.py [hunk 7]
```diff
@@ -298,8 +342,118 @@ class Charge(StripeObject):
...
+    last4 = models.CharField(max_length=4)
+    metadata = JSONField(null=True)
+    routing_number = models.TextField()
+    status = models.TextField()
+
+    @property
+    def stripe_bankaccount(self):
+        return self.account.stripe_account.external_accounts.retrieve(
+            self.stripe_id
+        )
```
Comment (blueyed) [inline pinax/stripe/models.py]: cached property?!

File: tox.ini [hunk 8]
```diff
@@ -30,4 +30,4 @@ setenv =
    LC_ALL=en_US.UTF-8
 commands =
     flake8 pinax
-    coverage run setup.py test
+    coverage run setup.py test {posargs}
```
Comment (blueyed) [inline tox.ini]: Added to https://github.com/pinax/pinax-stripe/pull/355/commits/f717fbee62606415550defdc16038791c7ac6308.

File: setup.py [hunk 9]
```diff
@@ -88,6 +88,7 @@
         "django>=1.7",
         "pytz",
         "six",
+        "django-ipware==1.1.6"
```
Comment (blueyed) [inline setup.py]: django-ipware is only used for the `InitialManagedAccountForm` - might be better to make this an optional dependency then only (in case it cannot be avoided altogether).

**Comment by paltman** [comment]
I starting to catch up on all the details (and trying to do so in the abstract as I've never dealt with connected accounts before).  At one point in this thread, it's mentioned that this is for `custom` type accounts.  Would that mean that `express` is inherently supported as a subset of what `custom` would require?

File: pinax/stripe/actions/charges.py [hunk 10]
```diff
@@ -61,20 +64,36 @@ def create(amount, customer, source=None, currency="usd", description=None, send
...
         source=source,
         customer=customer,
         description=description,
-        capture=capture,
+        capture=capture
+    )
+    if stripe_account:
+        kwargs['destination'] = {
+            'account': stripe_account
+        }
```
Comment (blueyed) [inline pinax/stripe/actions/charges.py]: Should there be support for direct charges, too?
https://stripe.com/docs/connect/charges#choosing-approach

Comment (lukeburden) [reply pinax/stripe/actions/charges.py]: @blueyed I can adjust this to allow for both direct and destination charges.

File: setup.py (hunk 9 above)
Comment (lukeburden) [reply setup.py]: @blueyed I've had a think about this - it strikes me that having an optional dependency creates complexity during deployment.

If you have a project that uses pinax-stripe and you are using InitialManagedAccountForm and installing pinax-stripe doesn't include django-ipware, doesn't that put the deployer in an odd situation where they have to remember to install django-ipware as part of their projects specific dependencies?

Let me know if you have a good reference implementation of an optional dependency in Python that I can take a peek at.

**Comment by lukeburden** [comment]
Hey crew,

An update on this. I've just rebased off master and will do the following as soon as I can:

- add basic support for direct, destination and separate charges. Nothing too fancy though.
- look at making django-ipware an optional dependency, although my sense is this just creates complication at the deployment phase (and thoughts there welcome)
- split out improvements to data syncing into a separate PR if it's not too much of a PITA
- basic tests for the dynamic account form, that self-configures on the basis of the specific verification required for a particular account.

@paltman - to answer your question, standard and express accounts will be support in so far as you can create a charge and your data will be sync'd properly - however, those both require integration with an oauth-like-flow that takes users to a Stripe hosted setup page for account creation. I have not added support for this flow in the form of a redirect view and callback view as I simply do not need this functionality for my use case. It shouldn't be too hard for someone else to add when they need it. :)

Two other issues:

- Stripe have [split payouts and transfers](https://stripe.com/docs/transfer-payout-split). I will not address this in this PR, and will leave adding support for this split for another PR.
- Stripe have [renamed "standalone" and "managed" accounts](https://stripe.com/docs/connect/custom-accounts#name-change) to "standard", "express" and "custom" accounts. I will also likely leave this for an extra PR. I'm worried adding support for this now will, by virtue of requiring us to default to a very recent version of Stripe's API, also force us to add support for the payout/transfer split.

Your thoughts are welcome - keen to get this code merged!

Luke

**Comment by paltman** [comment]
Awesome work!  Excited about this!

giphy 7

**Comment by blueyed** [comment]
@paltman 
I would prefer to not have gifs in here.. :/

**Comment by blueyed** [comment]
@lukeburden 
Could you make use of my branch with fixups?

**Comment by blueyed** [comment]
@lukeburden 

> Stripe have renamed "standalone" and "managed" accounts to "standard", "express" and "custom" accounts. I will also likely leave this for an extra PR. I'm worried adding support for this now will, by virtue of requiring us to default to a very recent version of Stripe's API, also force us to add support for the payout/transfer split.

It seems more that this renaming should get reflected in pinax-stripe after all.
There does not seem to be an API breakage involved here, but it seems to be rather about how things are called, and therefore it makes sense to include it here.

**Comment by blueyed** [comment]
> split out improvements to data syncing into a separate PR if it's not too much of a PITA

I can help with that (do not remember if I've done it already).

File: setup.py (hunk 9 above)
Comment (blueyed) [reply setup.py]: You can use `extras_require` for this, see e.g. https://github.com/celery/celery/blob/de2d075fb07850eee247766ebeb11d5530a43734/setup.py#L217 for something more involved.

Comment (blueyed) [reply setup.py]: There could then be e.g. `pinax-stripe[forms]` (to have all dependencies for forms) and/or `pinax-stripe[full]` (to really have all optional deps).

For now `pinax-stripe[ipware]` could be used - but I am not sure how much hassle it is for when you expect that everything works out of the box?!

File: pinax/stripe/models.py (hunk 6 above)
Comment (lukeburden) [reply pinax/stripe/models.py]: @blueyed - that hash you reference doesn't seem to link to a relevant change.

I'm going to leave these as properties for now, as this is what is used in general for accessing the Stripe object related to local instances. If we suddenly cache this, we could break people's code in subtle ways. They may, for example, call it multiple times on an instance and expect to see changes reflected in the returned value.

File: setup.py (hunk 9 above)
Comment (lukeburden) [reply setup.py]: Neat that there is a mechanism for this.

If `django-ipware` was a heavyweight dependency with serious storage, memory or cascading dependency consequences it would seem worthwhile to me to go that route. As it stands, `django-ipware` is a lightweight, pure-python lib with no dependencies. My gut instinct is that it is not a strong candidate for use of `extras_require`.

Comment (blueyed) [reply setup.py]: > is a lightweight

Even though it adds up in your requirements - and I assume thata lot of users of this lib are not using forms (and even this specific) one.

**Comment by blueyed** [comment]
Let's push this, but alsoo have some still:

image

File: setup.py (hunk 9 above)
Comment (lukeburden) [reply setup.py]: @blueyed once we're merged, feel free to knock up a PR and we can discuss with @paltman. My sense is that adding complexity to avoid a tiny dependency is not the way to go so I'm not keen on spending time further time on this.

File: pinax/stripe/actions/subscriptions.py [hunk 11]
```diff
@@ -117,24 +116,20 @@ def retrieve(customer, sub_id):
...
     if not sub_id:
         return
-    try:
-        return customer.stripe_customer.subscriptions.retrieve(sub_id)
-    except stripe.InvalidRequestError as e:
-        if smart_str(e).find("does not have a subscription with ID") == -1:
-            raise
+    subscription = stripe.Subscription.retrieve(sub_id)
+    # if subscription and not subscription.customer == customer.stripe_id:
+    #     return
```
[the remark this answers is not in the document]
Comment (lukeburden) [reply pinax/stripe/actions/subscriptions.py]: I'll just uncomment this for now, leaving the behaviour unchanged.

File: pinax/stripe/models.py (hunk 5 above)
Comment (lukeburden) [reply pinax/stripe/models.py]: Care to explain why those more verbose approaches would be better? :)

File: pinax/stripe/actions/charges.py (hunk 10 above)
Comment (lukeburden) [reply pinax/stripe/actions/charges.py]: I've now done this - take a peek!

**Comment by lukeburden** [comment]
I've spent some time on this today:

- removed changes to sync_customers, reverting to that in master (I have some neat code that I'd like to put into an improved version still, at some stage! Syncing using the current command on a modest project would take many days of runtime)
- updated code and migrations to reflect the current naming conventions for Connect (standard, express and custom accounts). I updated the names of the forms too.
- tested the InitialCustomAccountForm with the prompting of @sreyemnayr 
- fix a subscription retrieval test
- changed all single quotes to the project standard, double quotes
- add support for destination, direct and flex Charge creation

Any other blockers before we merge this beast?

**Comment by paltman** [comment]
@lukeburden this is a tremendous amount of quality work.  I really appreciate it.  I'll dig in next week and start reviewing things closer and figuring out how to test things out locally.  Any guidance you could provide would be helpful.  In fact, a short bit in the docs about this feature would be greatly appreciated, while it's all still fresh in your mind.  If you are not up for it, I completely understand and I'm happy to write the docs if you can provide me some bullet points / key points / outline and use that combined with some testing direction, I can write out the docs.

The only other thing I can think of prior to doing a more intensive review is just considering more tests.  We have maintained 100% coverage and I'd like to continue that.  I can help with tests as I have time but you probably have much more context to write better tests, than I.  But I'll want to get this branch up to 100% before merging.

I feel really bad asking even more of you (docs and tests) after all this work because I really do appreciate all the work you've done so far and am excited to get this merged in.  I'll do what I can to help with these requests.

**Comment by lukeburden** [comment]
Most of this is not too dissimilar to the rest of the project, I would say. In order to properly use custom accounts, people will need to write their own views and subclass the forms provided. They'll also need to hook up their own logic to webhooks for when custom accounts require further verification. I can definitely write a little in the docs about how this works.

If you have time @paltman it'd be great if you helped with some tests. I can write some too. Maybe @blueyed, if you have time, could pitch in a little? Your feedback so far has been great.

Also @blueyed - I saw you had some minor edits on a branch in your own fork. Would you be able to squash them and submit a PR against the `connect` branch on my fork? (my fork is still called django-stripe-payment).

**Comment by lukeburden** [comment]
I'll try put some time into this some evening this coming week. Looking forward to your review, @paltman!

**Comment by blueyed** [comment]
@lukeburden 
> I saw you had some minor edits on a branch in your own fork. Would you be able to squash them and submit a PR against the connect branch on my fork?

I've notified you about them 2 weeks ago (https://github.com/pinax/pinax-stripe/pull/336#pullrequestreview-60393191), and later asked about it again (https://github.com/pinax/pinax-stripe/pull/336#issuecomment-329295252).
They were meant to be squashed into your existing commits - and I've reordered/applied them accordingly already.

Since you have force-pushed your branch (eaaac17...a3f7f87) it causes conflicts now.
And I've noticed that there are still commits that have conflict markers in your branch anyway.
So I will rather wait for feedback from you on how to proceed here.  I think the best would be to apply the patches from my branch still, which could be done during an interactive rebase on your side.
You can see them here: https://github.com/blueyed/pinax-stripe/compare/eaaac17...connect-fixes

**Comment by lukeburden** [comment]
@blueyed - I'd rather you stuck to the regular workflow on Github and submitted a PR. If you had provided a PR in the first instance, I'd have reviewed and merged. The force update was an oversight on my side; not the best move for collaboration - noted and I'll avoid in the future.

For your PR, I'd recommend cherry-picking your specific fix commits onto a branch based off of the HEAD of the branch behind this PR.

**Comment by blueyed** [comment]
> I'd rather you stuck to the regular workflow on Github and submitted a PR

It is not common to have PRs for PRs, is it for you?

> I'd have reviewed and merged

And after all you cannot do that on GitHub with fixup commits.

> I'd recommend cherry-picking your specific fix commits onto a branch based off of the HEAD of the branch behind this PR.

My commits are not on top of your (original) branch/PR but meant to be squashed/fixed up in there - I've gone through it with interactive rebase and amended stuff.

**Comment by lukeburden** [comment]
@blueyed - your help is truly appreciated - I don't mean to be a pain, I just don't have the time to play around with your branch and commits. Thanks! Got to switch focus back to work now. I'll await that PR.

**Comment by blueyed** [comment]
All good.

> I'll await that PR.

Don't wait for it though - please just go ahead.

**Comment by paltman** [comment]
> I think that @paltman has no problems with merging commits with conflict makers

I'm not sure what is meant by _conflict makers_.  If that was a typo and you meant _conflict markers_ then I think you previously misunderstood me.  I'd prefer that those be resolved before pushing.  I don't think having:

```diff
+++
===
---
```

lines in the middle of a Python file helps anyoone.

**Comment by blueyed** [comment]
@paltman 
Yeah.. markers of course..
but they are in here (in the commit history) and my impressions was that you would not care as long as the final diff does not contain them.

**Comment by blueyed** [comment]
(as for having them in the Git history this is bad with using git-bisect, and even empty or non-squashed commits are bad then (more overhead))

**Comment by paltman** [comment]
Ah, then it was my misunderstanding.  I merged based on the fact that it was a PR without a WIP message or title and the resulting diff looked great.  It's there now.  Nothing we can really do about it.

**Comment by lukeburden** [comment]
I've added some tests for the Account forms - a little work there to do yet, namely testing an ID document upload. Coverage should almost be there.

A little splash of documentation and we should be close to sweet, sweet merge time!

**Comment by lukeburden** [comment]
@paltman - I've pushed coverage about as far as I have time for. Going to turn to the documentation next!

**Comment by blueyed** [comment]
Can we get https://github.com/pinax/pinax-stripe/pull/373 merged first, to have a better overview about coverage (changes)?

**Comment by paltman** [comment]
Fantastic. I’ll check it out.

File: pinax/stripe/actions/accounts.py [hunk 12]
```diff
@@ -1,9 +1,8 @@
-import stripe
-
+from . externalaccounts import sync_bank_account_from_stripe_data
 from .. import models
 from .. import utils
 import datetime
-from . externalaccounts import sync_bank_account_from_stripe_data
+import stripe
```
Comment (paltman) [inline pinax/stripe/actions/accounts.py]: Noticing some of the import ordering change around.  What's the reason for it?

In Pinax, we structure the imports like so:

```python
# first set of imports are stdlib imports
# non-from imports go first then from style import in their own group
import csv

# second set of imports are Django imports with contrib in their own
# group.
from django.urls import reverse
from django.db import models
from django.utils import timezone
from django.utils.translation import ugettext_lazy as _

from django.contrib.auth.models import User

# third set of imports are external apps (if applicable)
from tagging.fields import TagField

# fourth set of imports are local apps
from .fields import MarkupField
```

Details can be found in [CONTRIBUTING.md](https://github.com/pinax/pinax-stripe/blob/master/CONTRIBUTING.md).  I'm happy to fix them all up because I know at this point you just want to be done :).

Mostly curious as to why you wanted to make the change.

Comment (lukeburden) [reply pinax/stripe/actions/accounts.py]: Ah, sorry - I am in the habit of alphabetically ordering them and being done with it. I find this to be the lowest friction approach. But I should have stuck to your standard - thanks for patching that up.

Comment (blueyed) [reply pinax/stripe/actions/accounts.py]: We (and Pinax in general) should have an isort config then (https://github.com/timothycrosley/isort).  There would be also flake8-isort to check them then, but the main benefit is that you can just `isort` yourself locally.

**Comment by paltman** [comment]
@lukeburden I'm sending you another PR shortly for the `master` failures.

How are the docs coming?

**Comment by lukeburden** [comment]
@paltman the docs are on my mind, just a bit slammed with my work so struggling to get to it. How would you recommend I add to them? What editor, or workflow do you use? How can I test what they'll look like on RTD, etc? A little info would be great to speed me up when I get to it.

**Comment by paltman** [comment]
Just markdown, don't stress about formatting.  Maybe just add a new markdown file in the docs/ folder.  I can tidy up formatting, linking, etc.  I'll test the docs myself doing a trial run setup and polish off the rough corners as well.  Just need the broad strokes.  I appreciate the patience on the back and forth of this really long PR.

**Comment by paltman** [comment]
I’ll merge this when I some free time this evening. 

Thanks for all this great work.
